File: 2023/11_CosmicExpansion/image.py

# ======================================================================
# Cosmic Expansion
#   Advent of Code 2023 Day 11 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         i m a g e . p y
# ======================================================================
"Image for the Advent of Code 2023 Day 11 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
SPACE = "."
GALAXY = "#"

# ======================================================================
#                                                                  Image
# ======================================================================


class Image(object):   # pylint: disable=R0902, R0903, R0205
    "Object for Cosmic Expansion"

    def __init__(self, text=None, part2=False, expansion=None):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.expand_col = []
        self.expand_row = []
        self.galaxies = []
        self.expansion = expansion
        if expansion is None:
            self.expansion = 1000000 if part2 else 2

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            self._process_text(text)

    def _process_text(self, text):
        "Assign values from text"

        # 0. Precondition axions
        assert text is not None and len(text) > 0

        # 1. Find the expanded column numbers
        self.expand_col = [0]
        expanded = 0
        for indx in range(len(self.text[0])):
            if self.is_col_empty(indx):
                expanded += self.expansion
            else:
                expanded += 1
            self.expand_col.append(expanded)

        # 2. Find the expanded row numbers
        self.expand_row = [0]
        expanded = 0
        for indx in range(len(self.text[0])):
            if self.is_row_empty(indx):
                expanded += self.expansion
            else:
                expanded += 1
            self.expand_row.append(expanded)

        # 3. Get the expanded location of the qalaxies
        for rindx, row in enumerate(self.text):
            for cindx, square in enumerate(row):
                if square == GALAXY:
                    exp_loc = (self.expand_row[rindx], self.expand_col[cindx])
                    self.galaxies.append(exp_loc)

    def is_row_empty(self, indx):
        "Return True if there are no galaxies in this row"

        return GALAXY not in self.text[indx]

    def is_col_empty(self, indx):
        "Return True if there are no galaxies in this column"

        return all(row[indx] == SPACE for _, row in enumerate(self.text))

    @staticmethod
    def manhattan_distance(loc1, loc2):
        "Returns the manhattan (taxi cab) distance between two locations"
        return abs(loc1[0] - loc2[0]) + abs(loc1[1] - loc2[1])

    def all_distances(self):
        "Returns the sum of all distances between pairs of galaxies"

        # 1. Start with nothing
        result = 0

        # 2. Loop for all pairs of galixies
        for indx1, gal1 in enumerate(self.galaxies):
            for indx2, gal2 in enumerate(self.galaxies):
                if indx1 <= indx2:
                    continue

                # 3. Determine the distance between the pair
                dist = Image.manhattan_distance(gal1, gal2)

                # 4. Accumulate the distances
                result += dist

        # 5. Return the accumulated distances
        return result

# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------


if __name__ == '__main__':
    pass

# ======================================================================
# end                         i m a g e . p y                        end
# ======================================================================

File: 2023/11_CosmicExpansion/image.go
// ======================================================================
// Cosmic Expansion
//   Advent of Code 2023 Day 11 -- Eric Wastl -- https://adventofcode.com
//
// go implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           i m a g e . g o
// ======================================================================
// A solver for the Advent of Code 2023 Day 11 puzzle

package main

// ----------------------------------------------------------------------
//                                                                imports
// ----------------------------------------------------------------------
import (
	"fmt"
	"strings"
)

// ----------------------------------------------------------------------
//
//	constants
//
// ----------------------------------------------------------------------
const (
	galaxy = '#'
)

// ======================================================================
//                                                                  Image
// ======================================================================

// Image ... Object for Cosmic Expansion
type Image struct {
	Part2    bool
	Text     []string
	Galaxies []*Galaxy
}

// NewImage ... Create a Image object
func NewImage(part2 bool, text []string) (*Image, error) {

	// 1. Set the initial values
	p := new(Image)
	p.Part2 = part2
	p.Text = text

	// 2. Process text (if any)
	if len(p.Text) > 0 {
		err := p.processText()
		if err != nil {
			return nil, err
		}
	}

	// 3. Return the new Image object
	return p, nil
}

// processText ... Assign values from text
func (p *Image) processText() error {

	// 1. Get the expansion increment
	expInc := int64(2)
	if p.Part2 {
		expInc = int64(1000000)
	}

	// 2. Get the expanded index for the columns
	expCols := make([]int64, 0, len(p.Text[0])+1)
	expandIndx := int64(0)
	expCols = append(expCols, expandIndx)
	for indx := 0; indx < len(p.Text[0]); indx = indx + 1 {
		if p.IsColEmpty(indx) {
			expandIndx = expandIndx + expInc
		} else {
			expandIndx = expandIndx + 1
		}
		expCols = append(expCols, expandIndx)
	}

	// 3. Get the expanded index for the rows
	expRows := make([]int64, 0, len(p.Text)+1)
	expandIndx = int64(0)
	expRows = append(expRows, expandIndx)
	for indx := 0; indx < len(p.Text); indx = indx + 1 {
		if p.IsRowEmpty(indx) {
			expandIndx = expandIndx + expInc
		} else {
			expandIndx = expandIndx + 1
		}
		expRows = append(expRows, expandIndx)
	}

	// 4. Loop through the image
	for rowIndx, row := range p.Text {
		for colIndx, col := range row {

			// 5. If there is a galaxy here, add it
			if col == galaxy {
				p.Galaxies = append(p.Galaxies, NewGalaxy(expRows[rowIndx], expCols[colIndx]))
			}
		}
	}

	// 5. Return success
	return nil
}

// IsRowEmpty ... Returns true if row is empty
func (p *Image) IsRowEmpty(indx int) bool {
	return !strings.Contains(p.Text[indx], string(galaxy))
}

// IsColEmpty ... Returns true if column is empty
func (p *Image) IsColEmpty(indx int) bool {

	// 1. Loop for all the rows
	for _, row := range p.Text {

		// 2. If this row has a galaxy in the column, return false
		if galaxy == row[indx] {
			return false
		}
	}

	// 3. Never did see a galaxy, return true
	return true
}

// DistanceAll ... Returns the sum of the distances between each galaxies
func (p *Image) DistanceAll() int64 {

	// 1. Start with nothing
	result := int64(0)

	// 2. Loop for each galaxy
	for indx, g := range p.Galaxies {

		// 3. Get the distances to each of the other new galaxies
		dist := g.DistanceAll(p.Galaxies[indx+1:])

		// 4. Accumulate the distances
		result = result + dist
	}

	// 5. Return the total of the distances
	return result
}

// PartOne ... Returns the solution for part one
func (p *Image) PartOne(verbose bool, limit int) string {

	// 1. Return the solution for part one
	return fmt.Sprint(p.DistanceAll())
}

// PartTwo ... Returns the solution for part two
func (p *Image) PartTwo(verbose bool, limit int) string {

	// 1. Return the solution for part two
	return fmt.Sprint(p.DistanceAll())
}

// ======================================================================
// end                         i m a g e . g o                        end
// ======================================================================
